subsys/canbus/canopen/CO_driver.c

```diff
@@ -0,0 +1,488 @@
+		CO_errorReport(CANmodule->em, CO_EM_MEMORY_ALLOCATION_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return CO_ERROR_OUT_OF_MEMORY;
+	}
+
+	return CO_ERROR_NO;
+}
+
+CO_CANtx_t *CO_CANtxBufferInit(CO_CANmodule_t *CANmodule, u16_t index,
+			       u16_t ident, bool_t rtr, u8_t noOfBytes,
+			       bool_t syncFlag)
+{
+	CO_CANtx_t *buffer;
+
+	if (!CANmodule || (index >= CANmodule->tx_size)) {
+		LOG_ERR("failed to initialize CAN rx buffer, illegal argument");
+		CO_errorReport(CANmodule->em, CO_EM_GENERIC_SOFTWARE_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return NULL;
+	}
+
+	buffer = &CANmodule->tx_array[index];
+	buffer->ident = ident;
+	buffer->rtr = rtr;
+	buffer->DLC = noOfBytes;
+	buffer->bufferFull = false;
+	buffer->syncFlag = syncFlag;
+
+	return buffer;
+}
+
+CO_ReturnError_t CO_CANsend(CO_CANmodule_t *CANmodule, CO_CANtx_t *buffer)
+{
+	CO_ReturnError_t ret = CO_ERROR_NO;
+	struct zcan_frame msg;
+	int err;
+
+	if (!CANmodule || !CANmodule->dev || !buffer) {
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	CO_LOCK_CAN_SEND();
+
+	if (buffer->bufferFull) {
+		if (!CANmodule->first_tx_msg) {
+			CO_errorReport(CANmodule->em, CO_EM_CAN_TX_OVERFLOW,
+				       CO_EMC_CAN_OVERRUN, buffer->ident);
+		}
+		buffer->bufferFull = false;
+		ret = CO_ERROR_TX_OVERFLOW;
+	}
+
+	msg.id_type = CAN_STANDARD_IDENTIFIER;
+	msg.std_id = buffer->ident;
+	msg.dlc = buffer->DLC;
+	msg.rtr = (buffer->rtr ? 1 : 0);
+	memcpy(msg.data, buffer->data, buffer->DLC);
+
+	err = can_send(CANmodule->dev, &msg, K_NO_WAIT, canopen_tx_isr_callback,
+		       CANmodule);
+	if (err == CAN_TIMEOUT) {
+		buffer->bufferFull = true;
+	} else if (err != CAN_TX_OK) {
+		LOG_ERR("failed to send CAN frame (err %d)", err);
+		CO_errorReport(CANmodule->em, CO_EM_GENERIC_SOFTWARE_ERROR,
+			       CO_EMC_COMMUNICATION, 0);
+		ret = CO_ERROR_TX_UNCONFIGURED;
+	}
+
+	CO_UNLOCK_CAN_SEND();
+
+	return ret;
+}
+
+void CO_CANclearPendingSyncPDOs(CO_CANmodule_t *CANmodule)
+{
+	bool_t tpdoDeleted = false;
+	CO_CANtx_t *buffer;
+	u16_t i;
+
+	if (!CANmodule) {
+		return;
+	}
+
+	CO_LOCK_CAN_SEND();
+
+	for (i = 0; i < CANmodule->tx_size; i++) {
+		buffer = &CANmodule->tx_array[i];
+		if (buffer->bufferFull && buffer->syncFlag) {
+			buffer->bufferFull = false;
+			tpdoDeleted = true;
+		}
+	}
+
+	CO_UNLOCK_CAN_SEND();
+
+	if (tpdoDeleted) {
+		CO_errorReport(CANmodule->em, CO_EM_TPDO_OUTSIDE_WINDOW,
+			       CO_EMC_COMMUNICATION, 0);
+	}
+}
+
+void CO_CANverifyErrors(CO_CANmodule_t *CANmodule)
+{
+	CO_EM_t *em = (CO_EM_t *)CANmodule->em;
+	struct can_bus_err_cnt err_cnt;
+	u8_t rx_overflows;
+	u32_t errors;
+
+	/*
+	 * TODO: Zephyr lacks an API for reading the rx mailbox
+	 * overflow counter.
+	 */
+	rx_overflows  = 0;
+
+	can_get_state(CANmodule->dev, &err_cnt);
+
+	errors = ((u32_t)err_cnt.tx_err_cnt << 16) |
+		 ((u32_t)err_cnt.rx_err_cnt << 8) |
+		 rx_overflows;
+
+	if (errors != CANmodule->errors) {
+		CANmodule->errors = errors;
+
+		if (err_cnt.tx_err_cnt >= 256U) {
```

Comment:
`err_cnt.tx_err_cnt >= 256U` won't ever evaluate to true. The counter only has 8 bits.
You can use the state return value from `can_get_state` anyway.

Reply:
@alexanderwachter Fixed, thanks.